source/dub/project.d

```diff
@@ -1604,16 +1624,23 @@ final class SelectedVersions {
 
 	static Json dependencyToJson(Dependency d)
 	{
-		if (d.path.empty) return Json(d.version_.toString());
+		if (!d.repository.empty) {
+			return serializeToJson([
+				"version": d.version_.toString(),
+				"repository": d.repository.remote,
+			]);
+		} else if (d.path.empty) return Json(d.version_.toString());
 		else return serializeToJson(["path": d.path.toString()]);
 	}
 
 	static Dependency dependencyFromJson(Json j)
 	{
 		if (j.type == Json.Type.string)
 			return Dependency(Version(j.get!string));
-		else if (j.type == Json.Type.object)
+		else if (j.type == Json.Type.object && "path" in j)
 			return Dependency(NativePath(j["path"].get!string));
+		else if (j.type == Json.Type.object && "repository" in j)
+			return Dependency(Repository(j["repository"].get!string), j["version"].get!string);
 		else throw new Exception(format("Unexpected type for dependency: %s", j.type));
```

Comment:
The error message can now print "object" as unexpected type if neither path or repository are specified. It should probably just print the json at this point to avoid confusion.